asapdiscovery-docking/asapdiscovery/docking/schema/pose_generation.py

```diff
@@ -0,0 +1,198 @@
+        reference_ligand: ReferenceLigand,
+        ligands: list[Ligand],
+        core_smarts: str,
+    ) -> list[Ligand]:
+        """The main worker method which should generate ligand poses in the receptor using the reference ligand where required."""
+        ...
+
+    def _validate_ligands(self, ligands: list[Ligand]):
+        """
+        For the given set of ligands make sure that the docked ligand is the intended ligand target i.e does the
+        3D stereo match what we intend at input.
+        """
+        pass
+
+    def generate_poses(
+        self,
+        receptor: PreppedTarget,
+        reference_ligand: ReferenceLigand,
+        ligands: list[Ligand],
+        core_smarts: str,
+    ) -> list[Ligand]:
+        """
+        Generate poses for the given list of molecules in the target receptor.
+
+        Note:
+            We assume all stereo and states have been expanded and checked by this point.
+
+        Parameters
+        ----------
+        receptor: The prepared receptor file with the reference ligand removed.
+        reference_ligand: The reference ligand which should be used to constrain the pose generation.
+        ligands: The list of ligands which require poses in the target receptor.
+        core_smarts: The smarts string which should be used to identify the MCS between the ligand and the reference.
+
+        Returns
+        -------
+            A list of ligands with new poses generated
+            # TODO maybe we need some pose generation result object? or to store the results on the ligand schema?
+        """
+        ligands = self._generate_poses(
+            receptor=receptor,
+            reference_ligand=reference_ligand,
+            ligands=ligands,
+            core_smarts=core_smarts,
+        )
+        self._validate_ligands(ligands=ligands)
+        return ligands
+
+
+class OpenEyeConstrainedPoseGenerator(_BasicConstrainedPoseGenerator):
+    type: Literal["OpenEyeConstrainedPoseGenerator"] = "OpenEyeConstrainedPoseGenerator"
+    max_confs: PositiveInt = Field(
+        1000, description="The maximum number of conformers to try and generate."
+    )
+    energy_window: PositiveFloat = Field(
+        20,
+        description="Sets the maximum allowable energy difference between the lowest and the highest energy conformers, in units of kcal/mol",
+    )
+    clash_cutoff: PositiveFloat = Field(2.0, description="The ")
+
+    def provenance(self) -> dict[str, Any]:
+        from openeye import oechem, oeomega
+
+        return {
+            "oechem": oechem.OEChemGetVersion(),
+            "oeomega": oeomega.OEOmegaGetVersion(),
+        }
+
+    def _generate_poses(
+        self,
+        receptor: PreppedTarget,
+        reference_ligand: ReferenceLigand,
+        ligands: list[Ligand],
+        core_smarts: str,
+    ) -> list[Ligand]:
+        """Use openeye oeomega to generate constrained poses for ligands"""
+
+        from openeye import oechem, oeomega
```

Comment:
We mask the import to check for a license
```suggestion
        asapdiscovery.data.openeye import oechem, oeomega
```

I might also suggest moving the core OE functionality to a function in the openeye module but up to you I know this is a draft.

Reply:
Thanks, I wasn't sure why this was done but makes sense!